File: java-base/src/main/java/net/cryptic_game/backend/base/api/ApiHandler.java
package net.cryptic_game.backend.base.api;

import java.util.HashMap;
import java.util.Map;

public class ApiHandler {

    private final Map<String, ApiEndpointExecutor> executors;
    private final Class<? extends ApiEndpointExecutor> executorClass;

    public ApiHandler(Class<? extends ApiEndpointExecutor> executorClass) {
        this.executors = new HashMap<>();
        this.executorClass = executorClass;
    }

    public ApiEndpointExecutor getEndpointExecutor(String endpoint) {
        return executors.get(endpoint);
    }

    public void registerEndpoint(final ApiEndpointExecutor executor) {
        this.executors.put(executor.getName(), executor);
    }

    public void registerApiCollection(final ApiCollection collection) {
        collection.load(executorClass).forEach(this::registerEndpoint);
    }
}

File: java-base/src/main/kotlin/net/cryptic_game/backend/base/api/ApiHandler.kt
package net.cryptic_game.backend.base.api

import kotlin.reflect.KClass

class ApiHandler(private val executorClass: Class<out ApiEndpointExecutor>) {
    constructor(executorClass: KClass<out ApiEndpointExecutor>) : this(executorClass.java)

    private val executors: MutableMap<String, ApiEndpointExecutor> = mutableMapOf()

    fun getEndpointExecutor(endpoint: String): ApiEndpointExecutor? = executors[endpoint]

    fun registerEndpoint(executor: ApiEndpointExecutor) {
        executors[executor.getName()] = executor
    }

    fun registerApiCollection(collection: ApiCollection) =
            collection.load(executorClass).forEach { executor: ApiEndpointExecutor -> registerEndpoint(executor) }
}